bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java

```diff
@@ -452,31 +457,48 @@ void doCompactEntryLogs(double threshold) {
 
         final int numBuckets = 10;
         int[] entryLogUsageBuckets = new int[numBuckets];
+        int[] compactedBuckets = new int[numBuckets];
+
+        long start = System.currentTimeMillis();
+        long end = start;
+        long timeDiff = 0;
 
         for (EntryLogMetadata meta : logsToCompact) {
             int bucketIndex = Math.min(
                     numBuckets - 1,
                     (int) Math.ceil(meta.getUsage() * numBuckets));
             entryLogUsageBuckets[bucketIndex]++;
 
-            if (meta.getUsage() >= threshold) {
+            if (timeDiff < limit) {
+                end = System.currentTimeMillis();
+                timeDiff = end - start;
+            }
+            if (meta.getUsage() >= threshold || (limit > 0 && timeDiff > limit) || !running) {
+                // We allow the usage limit calculation to continue so that we get a accurate
+                // report of where the usage was prior to running compaction.
                 continue;
```

Comment:
We should break the loop if exceed the max compaction time? all the remaining logs will be skipped.

Reply:
The consumer, would not get any indication of how may logs were not compacted if we skip out of loop. This ensures that the exit report of the usage buckets give a complete picture of what was there and what was compacted. The usage gathering from what I saw as extremely fast compared to the compaction. The logs show that this does not take a appreciable amount of time.  However, a consumer whom is tuning the system may need this insight.